proposed/local-sdk-global-json.md

```diff
@@ -0,0 +1,215 @@
+
+This problem also manifests in how customers naturally want to use our
+development tools like Visual Studio or VS Code. It's felt sharply on the .NET
+team, or any external customer who wants to contribute to .NET, due to how
+arcade infrastructure uses xcopy deployment into `.dotnet`. External teams
+like Unity also feel this pain in their development:
+
+- This [issue][cases-sdk-issue] from 2017 attempting
+to solve this problsem. It gets several hits a year from customers who are
+similarly struggling with our toolings inability to handle local deployment.
+- This [internal discussion][cases-internal-discussion] from a C# team member.
+They wanted to use VS as the product is shipped to customers and got blocked
+when we shipped an SDK that didn't have a corresponding MSI and hence VS
+couldn't load Roslyn anymore.
+- [VS code][cases-vscode] having to adjust to consider local directories for SDK
+because our resolver can't find them.
+
+## Detailed Design
+
+The global.json file will support two new properties under the `sdk` object:
+
+- `"paths"`: this is a list of paths that the host resolver should
+consider when looking for compatible SDKs. In the case this property is `null`
+or not specified, the host resolver will behave as it does today.
+- `"errorMessage"`: when the host resolver cannot find a compatible .NET SDK it
+will display the contents of this property instead of the default error message.
+In the case this property is `null` or not specified, the current error message
+will be displayed.
+
+The values in the `paths` property can be a relative path, absolute path or
+`$host$`.  When a relative path is used it will be resolved relative to the
+location of the containing global.json. The value `$host$` is a special value
+that represents the machine wide installation path of .NET SDK for the
+[current host][installation-doc].
+
+The values in `paths` are considered in the order they are defined. The host
+resolver will stop when it finds the first path with a compatible .NET SDK.
+For example:
+
+```json
+{
+    "sdk": {
+        "paths": [ ".dotnet", "$host$" ],
+    }
+}
+```
+
+In this configuration the host resolver would find a compatible .NET SDK, if it
+exists in `.dotnet` or a machine wide location.
+
+This lookup will stop on the first match which means it won't necessarily find
+the best match. Consider a scenario with a global.json that has:
+
+```json
+{
+  "sdk": {
+    "paths": [ ".dotnet", "$host$" ],
+    "version": "7.0.200",
+    "rollForward": "latestFeature"
+  }
+}
+```
+
+In a scenario where the `.dotnet` directory had 7.0.200 SDK but there was a
+machine wide install of 7.0.300 SDK, the host resolver would pick 7.0.200 out
+of `.dotnet`. That location is considered first, it has a matching .NET SDK and
+hence discovery stops there.
+
+This design requires us to only change the host resolver. That means other
+tooling like Visual Studio, VS Code, MSBuild, etc ... would transparently
```

Comment:
I'll leave the detailed analysis to @elinor-fung , but I think it will work out that way. The changes will be in `hostfxr` and we've been advocating for all of the other tools to use `hostfxr` to locate the SDK. Specifically `MSBuildLocator` does use `hostfxr`. That said, I think it would be good to include testing of some of the obvious use cases as part of the feature work on this (we've been surprised in the past unfortunately).

Reply:
Additionally, the fact this lives in `hostfxr` will mean that it's versioning behavior will be somewhat "unexpected". As long as one has the .NET 9 version with this change installed, it will work regardless of the SDK version chosen in the `global.json` because we always use the latest `hostfxr` available in a given installation. This includes previews, so just installing a preview of .NET 9 will make this work basically everywhere on the machine.

Reply:
I will take this to mean VS will partially work, VS Code will work fully. As a VS Code user, I am fine with this 😄

Reply:
VS will work the way it always was - even today if you set `PATH` and `DOTNET_ROOT` which is what the various scripts in repos do, before starting VS, VS will still run the .NET Framework version of MSBuild it ships with for design builds and the other one for build. But it should still use `hostfxr` to find the SDK to use for the given app. So this change should make it unnecessary to use the special scripts - which is probably 80% of the value of this feature :-)
That is the one scenario we definitely need to test when developing this feature.

Reply:
...including the out-of-proc scenarios (e.g., the Windows Forms designers). /cc: @merriemcgaw

Reply:
> VS uses the .NET Framework version of MSBuild it ships with, period, for evaluation and for all builds including design time and F5 builds.

Thanks for the explanation!